activerecord/lib/active_record/associations/through_association.rb

```diff
@@ -71,7 +71,15 @@ def stale_state
         end
 
         def foreign_key_present?
-          through_reflection.belongs_to? && !owner[through_reflection.foreign_key].nil?
+          (through_reflection.belongs_to? && !owner[through_reflection.foreign_key].nil?) ||
+            recursive_through_foreign_key_present?(through_reflection.through_reflection)
```

Comment:
Couldn't this also work without the need for the second method, which is basically the same code, by calling `foreign_key_present?` recursively?

```
def foreign_key_present?
  (through_reflection.belongs_to? && !owner[through_reflection.foreign_key].nil?) ||
  through_reflection.through_reflection.try(:foreign_key_present?)
end
```